### PR TITLE
Update R3006,R2016

### PR DESCRIPTION
**Pull request by jianyexi**

### CODE REVIEW

File: src/dotnet/OpenAPI.Validator.Tests/OpenAPIModelerValidationTests.cs [hunk 1]
```diff
@@ -907,6 +915,13 @@ public void ValidXMSPathWithOAata()
             }
             Assert.Empty(message);
         }
+
+        [Fact]
+        public void ValidRequiredDesciminatorPropertiesInPatchRequestValidation()
```
Comment (lirenhe) [inline src/dotnet/OpenAPI.Validator.Tests/OpenAPIModelerValidationTests.cs]: do you have the existing sample to verify R2016? I don't think we need to create a new test case for this.

File: src/dotnet/OpenAPI.Validator/Model/Utilities/ValidationUtilities.cs [hunk 2]
```diff
@@ -169,6 +169,25 @@ public static IEnumerable<string> EnumerateRequiredProperties(string modelName,
...
+        /// <param name="modelName">model for which to check the discrimintor properties</param>
+        /// <param name="definitions">dictionary of model definitions</param>
+        /// <param name="propertyList">List of required properties found in model hierarchy</param>
+        public static IEnumerable<string> EnumerateDiscrimintorProperties(string modelName, Dictionary<string, Schema> definitions)
+        {
+            var modelsToCheck = EnumerateModelHierarchy(modelName, definitions);
+            var propertiesList = new List<string>();
+            foreach (var modelRef in modelsToCheck)
+            {
+                if (!definitions.ContainsKey(modelRef) || definitions[modelRef].Discriminator?.Any() != true) continue;
```
Comment (lirenhe) [inline src/dotnet/OpenAPI.Validator/Model/Utilities/ValidationUtilities.cs]: could you refactor this code?  no need to combine them in one line.

File: src/dotnet/OpenAPI.Validator/Model/Utilities/ValidationUtilities.cs [hunk 3]
```diff
@@ -169,6 +169,25 @@ public static IEnumerable<string> EnumerateRequiredProperties(string modelName,
             return propertiesList;
         }
 
+        /// <summary>
+        /// Returns the cumulative list of all discrimintor properties found in the entire model hierarchy
+        /// </summary>discrimintor
+        /// <param name="modelName">model for which to check the discrimintor properties</param>
+        /// <param name="definitions">dictionary of model definitions</param>
+        /// <param name="propertyList">List of required properties found in model hierarchy</param>
```
Comment (lirenhe) [inline src/dotnet/OpenAPI.Validator/Model/Utilities/ValidationUtilities.cs]: is this parameter used?

Comment (jianyexi) [reply src/dotnet/OpenAPI.Validator/Model/Utilities/ValidationUtilities.cs]: thank you, I will remove it

File: src/dotnet/OpenAPI.Validator.Tests/OpenAPIModelerValidationTests.cs (hunk 1 above)
Comment (jianyexi) [reply src/dotnet/OpenAPI.Validator.Tests/OpenAPIModelerValidationTests.cs]: this case is a litter special, so I add a new to test it
Btw I think we should modify one test case only when it has some bug

File: src/dotnet/OpenAPI.Validator/Model/Utilities/ValidationUtilities.cs (hunk 2 above)
Comment (jianyexi) [reply src/dotnet/OpenAPI.Validator/Model/Utilities/ValidationUtilities.cs]: this '!= true' can not be replaced by '!', or will compile failed

File: src/dotnet/OpenAPI.Validator.Tests/OpenAPIModelerValidationTests.cs (hunk 1 above)
Comment (lirenhe) [reply src/dotnet/OpenAPI.Validator.Tests/OpenAPIModelerValidationTests.cs]: Not necessary. But anyway, you could make the decision for this.

File: src/dotnet/OpenAPI.Validator/Model/Utilities/ValidationUtilities.cs [hunk 4]
```diff
@@ -169,6 +169,26 @@ public static IEnumerable<string> EnumerateRequiredProperties(string modelName,
...
+        /// </summary>discrimintor
+        /// <param name="modelName">model for which to check the discrimintor properties</param>
+        /// <param name="definitions">dictionary of model definitions</param>
+        public static IEnumerable<string> EnumerateDiscrimintorProperties(string modelName, Dictionary<string, Schema> definitions)
+        {
+            var modelsToCheck = EnumerateModelHierarchy(modelName, definitions);
+            var propertiesList = new List<string>();
+            foreach (var modelRef in modelsToCheck)
+            {
+                if (!definitions.ContainsKey(modelRef) || definitions[modelRef].Discriminator?.Any() != true) {
```
Comment (lirenhe) [inline src/dotnet/OpenAPI.Validator/Model/Utilities/ValidationUtilities.cs]: Could you use '!definitions[modelRef].Discriminator?.Any()' directly?

Comment (ruowan) [inline src/dotnet/OpenAPI.Validator/Model/Utilities/ValidationUtilities.cs]: Why use Any()? use String.IsNullOrEmpty(String)  is more clear for string type.

Comment (jianyexi) [reply src/dotnet/OpenAPI.Validator/Model/Utilities/ValidationUtilities.cs]: Good suggestion!

**Pull request merged**